ambry-tools/src/main/java/com.github.ambry.tools/perf/rest/NettyPerfClient.java

```diff
@@ -0,0 +1,501 @@
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private HttpRequest request;
+    private HttpResponse response;
+    private ChunkedInput<HttpContent> chunkedInput;
+
+    private int chunksReceived;
+    private long sizeReceived;
+    private long lastChunkReceiveTime;
+    private long requestStartTime;
+    private long requestId = 0;
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) {
+      perfClientMetrics.channelCreationRate.mark();
+      logger.trace("Channel {} active", ctx.channel());
+      sendRequest(ctx);
+    }
+
+    @Override
+    public void channelRead0(ChannelHandlerContext ctx, Object in) {
+      long currentChunkReceiveTime = System.currentTimeMillis();
+      boolean recognized = false;
+      if (in instanceof HttpResponse) {
+        recognized = true;
+        long responseReceiveStart = currentChunkReceiveTime - requestStartTime;
+        perfClientMetrics.timeToFirstResponseChunkInMs.update(responseReceiveStart);
+        logger.trace("Response receive has started on channel {}. Took {} ms", ctx.channel(), responseReceiveStart);
+        response = (HttpResponse) in;
+      }
+      if (in instanceof HttpContent) {
+        recognized = true;
+        perfClientMetrics.delayBetweenChunkReceiveInMs.update(currentChunkReceiveTime - lastChunkReceiveTime);
+        chunksReceived++;
+        int bytesReceivedThisTime = ((HttpContent) in).content().readableBytes();
+        sizeReceived += bytesReceivedThisTime;
+        perfClientMetrics.bytesReceiveRate.mark(bytesReceivedThisTime);
+        if (in instanceof LastHttpContent) {
+          long requestRoundTripTime = currentChunkReceiveTime - requestStartTime;
+          perfClientMetrics.requestRoundTripTimeInMs.update(requestRoundTripTime);
+          perfClientMetrics.getContentSizeInBytes.update(sizeReceived);
+          perfClientMetrics.getChunkCount.update(chunksReceived);
+          logger.trace(
+              "Final content received on channel {}. Took {} ms. Total chunks received - {}. Total size received - {}",
+              ctx.channel(), requestRoundTripTime, chunksReceived, sizeReceived);
+          if (HttpHeaders.isKeepAlive(response) && isRunning) {
+            logger.trace("Sending new request on channel {}", ctx.channel());
+            sendRequest(ctx);
+          } else if (!isRunning) {
+            logger.info("Closing channel {} because NettyPerfClient has been shutdown", ctx.channel());
+            ctx.close();
+          } else {
+            perfClientMetrics.requestResponseError.inc();
+            logger.error("Channel {} not kept alive. Last response status was {}", ctx.channel(), response.getStatus());
+            ctx.close();
+          }
+        }
+      }
+      if (!recognized) {
+        throw new IllegalStateException("Unexpected HttpObject - " + in.getClass());
+      }
+      lastChunkReceiveTime = currentChunkReceiveTime;
```

Comment:
If `recognized=false`, `lastChunkReceiveTime` will still get updated. Is this intended?